3scale/porta
# 🧠🌴 Fix Braintree 3DS in Dev portal (and much more)

### Description
[THREESCALE-8967: Adding Braintree SCA card fails](https://issues.redhat.com/browse/THREESCALE-8967)

The 3DS card verification is broken. It happens because we're sending `billingAddress: { region }` as a full name instead of a ISO 3166 code, i.e. `Malaga` vs. `MA`.

*Most importantly*, card verification via 3DS flow is being ignored. It was assumed that the `nonce` would not be given when card was invalid but this wasn't true, there is always a nonce. It's the shift in liability what has to be checked, and the status of the payload returned by `verifyCard`.

This PR addresses these problems and other potential pitfalls by fixing, improving and refactoring out `BraintreeForm` component used in the dev portal.

### Verification
Testing credit cards: https://developer.paypal.com/braintree/docs/guides/3d-secure/testing-go-live#testing

Assuming the payment gateway is set up as Braintree (sandbox account available in Bitwarden), setting `Finance` is enabled for the provider account and dev portal has feature `Finance` visible:

* For provider:
  1. Access your provider's payment details at `Settings > Billing`
  2. Add or edit the card card details

* For dev-portal:
  1. Mark *3D Secure enabled* at `Audience > Billing > Payment Gateway`
  2. Log in to dev portal go to `Settings > Credit card details > Add/Edit credit card details and billing address`
  3. Try to break the form in any way possible. The CC should only be saved if it passes 3DS verification. This modal should be required when a challenge is requested by the bank:
    <img width="600" alt="Screenshot 2023-02-17 at 08 49 20" src="https://user-images.githubusercontent.com/11672286/219583365-faee9e3a-9889-48b2-a493-c27ca5773a8b.png">

### Notes
Original PR:
* https://github.com/3scale/porta/pull/3079
